### ISSUE TITLE
Database update/migration failed from v12.9.2-0055 to v13.0.3-0055

### ISSUE THREAD

**Issue by helmut-steiner** [jboxberger/synology-gitlab#44]
Hey,
I ran the update from GitLab v12.9.2-0055 to v13.0.3-0055 early this morning but it seems like my database was not correctly migrated to the new postgresql version. After the update I was presented with a password reset screen not with my usual login screen and I even after that I couldn't login anymore with my username and password. 
I had to uninstall the new version, clear the files, reinstall the old version and reload my backups. It looked like it didn't port any of my settings over but the upgrade logs didn't show any abnormalities. 
Kind regards,
Helmut

**Comment by jboxberger** [comment]
Hi Helmut,

thank you for your Issue, I will investigate this update path. I am very glad that you had Backups and could get it working. Usually you can uninstall the broken version with the checkbox "keep data" checked and reinstall the previous version with the option use existing Data. But Backup is the best and cleanest way. Please keep this procedure in the future and backup before update. 

Do yoy have Complex CI Pipelines or big BLOBs ind you Gitlab Instance?

I will give you feedback asap.

**Comment by helmut-steiner** [comment]
Thanks for the swift reply! 
I do have a couple of CI pipelines, not sure about the BLOBs. Keeping the data during uninstallation didn't work as after reinstalling the old version it failed to start with an error code. It seemed as it altered the old database as well during the update process.

**Comment by helmut-steiner** [comment]
Also I still had the old database folder "10" next to the folder "12" and the upgrade log files (most of them empty or one liners) including a lock file in the postgresql folder. So I guess it must have crashed somewhere along the way without an error code.

**Comment by jboxberger** [comment]
Hello Helmut,

i've tested the Upgrade 2 or 3 times, but i test it with a limited set on my dev System. I could not find any issues. So i think there could be an issue with your data set.

Maybe you can check your LogFiles and compare them with mine.

synology_gitlab_postgresql: [synology_gitlab_postgresql.log](https://github.com/jboxberger/synology-gitlab/files/4742588/synology_gitlab_postgresql.log)
synology_gitlab: [synology_gitlab.log](https://github.com/jboxberger/synology-gitlab/files/4742589/synology_gitlab.log)

I've also build you one version with postgres:10 for testing (synology-gitlab-stock-aio-13.0.3-0055.zip) and an updated one so you can test maybe the newer version migrate without issues (synology-gitlab-stock-aio-13.0.5-0055.zip). Please note that i haven't tested the 13.0.5-0055 yet so handle with care and as always, backup, backup, backup :-).

The Download is valid for 7 Days (i recommend chromium for download): 
https://send.firefox.com/download/b37c3686135896ac/#hLz7dgkyAqrOytmbEc0R0w

Please let me know about your progress, would love to see your issue solved.

**Comment by helmut-steiner** [comment]
Thanks a lot for your tests and effort! I will give it a shot on the weekend. Can't risk having troubles in the production environment during work hours. :)
I will keep you in the loop if it works!

**Comment by iConMayrhofer** [comment]
Got the same error. Gitlab was installed like new. There where no configurations on the login screen which I made.
Also the log says: 

> 
Setting up GitLab for firstrun. Please be patient, this could take a while...

**Comment by jboxberger** [comment]
Thank you for your feedback. I've revoked the release. 

There seems to be a problem with the PostgreSQL migration routine within the PostgreSQL container... if you check my logs you will find at the migration at the bottom, could you please post your whole log.

My PostgreSQL Log.

2020-06-07 17:06:57	stdout	‣ Migration in progress. Please be patient...Performing Consistency Checks
2020-06-07 17:06:54	stdout	debconf: delaying package configuration, since apt-utils is not installed
2020-06-07 17:06:50	stdout	‣ Installing PostgreSQL 10...
2020-06-07 17:06:50	stdout	**‣ Migrating PostgreSQL 10 data to 12...**
2020-06-07 17:06:50	stdout	Initializing database...
2020-06-07 17:06:50	stdout	setfacl: /etc/resolv.conf: Operation not supported
2020-06-07 17:06:50	stdout	Setting resolv.conf ACLs...
2020-06-07 17:06:50	stdout	Initializing rundir...
2020-06-07 17:06:50	stdout	Initializing logdir...
2020-06-07 17:06:50	stdout	Initializing certdir...
2020-06-07 17:06:49	stdout	Initializing datadir..


Kind regards

**Comment by iConMayrhofer** [comment]
Here is my log

[synology_gitlab_postgresql.log](https://github.com/jboxberger/synology-gitlab/files/4748152/synology_gitlab_postgresql.log)

**Comment by iConMayrhofer** [comment]
Seems to have done the same, or?

**Comment by jboxberger** [comment]
Well not really.. your container is shutting down during migration...

```
mine:
2020-06-08 16:42:46	stdout	Initializing datadir...
2020-06-08 16:42:00	stdout	  gitlab
2020-06-08 16:41:52	stdout	Creating dump of database schemas
2020-06-08 16:41:52	stdout	Creating dump of global objects                             ok
2020-06-08 16:41:51	stdout	Checking for invalid "sql_identifier" user columns          ok
2020-06-08 16:41:51	stdout	Checking for tables WITH OIDS                               ok
2020-06-08 16:41:51	stdout	Checking for contrib/isn with bigint-passing mismatch       ok
2020-06-08 16:41:51	stdout	Checking for reg* data types in user tables                 ok
2020-06-08 16:41:51	stdout	Checking for prepared transactions                          ok
2020-06-08 16:41:51	stdout	Checking database connection settings                       ok
2020-06-08 16:41:51	stdout	Checking database user is the install user                  ok
2020-06-08 16:41:46	stdout	Checking cluster versions                                   ok
2020-06-08 16:41:46	stdout	-----------------------------
2020-06-08 16:41:46	stdout	‣ Migration in progress. Please be patient...Performing Consistency Checks
2020-06-08 16:40:48	stdout	debconf: delaying package configuration, since apt-utils is not installed
2020-06-08 16:40:29	stdout	‣ Installing PostgreSQL 10...
2020-06-08 16:40:29	stdout	‣ Migrating PostgreSQL 10 data to 12...
2020-06-08 16:40:29	stdout	Initializing database...
2020-06-08 16:40:29	stdout	setfacl: /etc/resolv.conf: Operation not supported
2020-06-08 16:40:29	stdout	Setting resolv.conf ACLs...
2020-06-08 16:40:29	stdout	Initializing rundir...
2020-06-08 16:40:29	stdout	Initializing logdir...
2020-06-08 16:40:29	stdout	Initializing certdir...
2020-06-08 16:40:25	stdout	Initializing datadir... 
```

```
yours
2020-06-07 17:07:51	stdout	Initializing datadir...
2020-06-07 17:07:45	stdout	2020-06-07 17:07:45.211 UTC [1] LOG:  **database system is shut** down
2020-06-07 17:07:43	stdout	2020-06-07 17:07:43.578 UTC [1665] LOG:  shutting down
2020-06-07 17:07:43	stdout	2020-06-07 17:07:43.577 UTC [1] LOG:  background worker "logical replication launcher" (PID 1670) exited with exit code 1
2020-06-07 17:07:43	stdout	2020-06-07 17:07:43.572 UTC [1] LOG:  received smart shutdown request
2020-06-07 17:07:03	stdout	2020-06-07 17:07:03.579 UTC [1] LOG:  database system is ready to accept connections
2020-06-07 17:07:03	stdout	2020-06-07 17:07:03.564 UTC [1664] LOG:  database system was shut down at 2020-06-07 17:07:03 UTC
2020-06-07 17:07:03	stdout	2020-06-07 17:07:03.545 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2020-06-07 17:07:03	stdout	2020-06-07 17:07:03.536 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2020-06-07 17:07:03	stdout	2020-06-07 17:07:03.536 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2020-06-07 17:07:03	stdout	2020-06-07 17:07:03.536 UTC [1] LOG:  starting PostgreSQL 12.3 (Ubuntu 12.3-1.pgdg18.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0, 64-bit
2020-06-07 17:07:03	stdout	Starting PostgreSQL 12...
2020-06-07 17:07:03	stdout	‣ Granting access to gitlab_user user...
2020-06-07 17:07:03	stdout	‣ Loading pg_trgm extension...
2020-06-07 17:07:03	stdout	Creating database: gitlab...
2020-06-07 17:07:03	stdout	Creating database user: gitlab_user
```

**Comment by iConMayrhofer** [comment]
Oh, you are right. Missed that

**Comment by jboxberger** [comment]
I try to find out why... i am using a vm on a ssd as test env. My System is much faster than a usual synology. After the installation the installation script shutdown all container and starts them again, this comes from the original synology package. maybe that's why the migration ist aborted by a container shutdown...

If this theory is correct, the synology-gitlab-stock-aio-13.0.3-0055.zip from here should work fine. It has the Postgres 10 bundled with the 13.0.3. If you feel comfortable with it you can give it a try. 
https://send.firefox.com/download/b37c3686135896ac/#hLz7dgkyAqrOytmbEc0R0w

i will check this. Thank you very much.

**Comment by iConMayrhofer** [comment]
I will first try to backup. If i have time after that, I'd be willing to try it.

**Comment by jboxberger** [comment]
sorry you sent me a too small snippet of your log so a compared the wrong places. could you please send me the full log.

**Comment by iConMayrhofer** [comment]
I did export this log from the Synology GUI. There wasn't more. Sorry

**Comment by jboxberger** [comment]
Do you see more in the GUI itself? Maybe the export function truncates something? 

In you log there are several Error from the migration itself. The Migration fails, so gitlabs strart with a clean db. But i can not see where the problems beginns because the log starts "in the middle" of the migration process.

```
2020-06-08 16:45:27	stdout	2020-06-08 16:45:27.647 UTC [177] STATEMENT:  INSERT INTO "application_settings" ("default_projects_limit", "signup_enabled", "gravatar_enabled", "created_at", "updated_at", "restricted_visibility_levels", "import_sources", "default_group_visibility", "repository_checks_enabled", "health_check_access_token", "repository_storages", "sign_in_text_html", "help_page_text_html", "shared_runners_text_html", "after_sign_up_text_html", "dsa_key_restriction", "plantuml_enabled", "unique_ips_limit_per_user", "unique_ips_limit_time_window", "default_artifacts_expire_in", "uuid", "cached_markdown_version", "password_authentication_enabled_for_web", "commit_email_hostname", "runners_registration_token_encrypted", "productivity_analytics_start_date", "id") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27) RETURNING "id"
2020-06-08 16:45:27	stdout	2020-06-08 16:45:27.647 UTC [177] DETAIL:  Key (id)=(1) already exists.
2020-06-08 16:45:27	stdout	2020-06-08 16:45:27.647 UTC [177] ERROR:  duplicate key value violates unique constraint "application_settings_pkey"
2020-06-08 16:43:59	stdout	2020-06-08 16:43:59.063 UTC [167] STATEMENT:  CREATE DATABASE "gitlab" ENCODING = 'unicode'
2020-06-08 16:43:59	stdout	2020-06-08 16:43:59.063 UTC [167] ERROR:  database "gitlab" already exists
2020-06-08 16:43:43	stdout	2020-06-08 16:43:43.997 UTC [162] STATEMENT:  SELECT "application_settings".* FROM "application_settings" ORDER BY "application_settings"."id" DESC LIMIT $1
2020-06-08 16:43:43	stdout	2020-06-08 16:43:43.997 UTC [162] ERROR:  relation "application_settings" does not exist at character 38
2020-06-08 16:43:43	stdout	2020-06-08 16:43:43.995 UTC [162] STATEMENT:  SELECT "application_settings".* FROM "application_settings" ORDER BY "application_settings"."id" DESC LIMIT $1
2020-06-08 16:43:43	stdout	2020-06-08 16:43:43.995 UTC [162] ERROR:  relation "application_settings" does not exist at character 38
2020-06-08 16:43:43	stdout	2020-06-08 16:43:43.891 UTC [161] STATEMENT:  SELECT "application_settings".* FROM "application_settings" ORDER BY "application_settings"."id" DESC LIMIT $1
2020-06-08 16:43:43	stdout	2020-06-08 16:43:43.891 UTC [161] ERROR:  relation "application_settings" does not exist at character 38
2020-06-08 16:43:43	stdout	2020-06-08 16:43:43.888 UTC [161] STATEMENT:  SELECT "application_settings".* FROM "application_settings" ORDER BY "application_settings"."id" DESC LIMIT $1
2020-06-08 16:43:43	stdout	2020-06-08 16:43:43.888 UTC [161] ERROR:  relation "application_settings" does not exist at character 38
2020-06-08 16:42:56	stdout	2020-06-08 16:42:56.557 UTC [1] LOG:  database system is ready to accept connections
```

**Comment by iConMayrhofer** [comment]
No there wasn't more. I also wondered why the log was so short. But I didn't have time to go into it in detail

**Comment by iConMayrhofer** [comment]
Maybe the GUI itself truncates the log? But that doesn't seem logical.
Where are this logs normally saved? I'm not that familar with docker.

**Comment by jboxberger** [comment]
No i meant the export maybe truncates or shortens it. iv'e copied mine direct from the GUI. Ok whene there is nothing more then i hope that helmut mybe will get some helpfull lines. As far as i can  see there ist something in the database that breaks the migration within the PSQL container. I dont think that i can solve the problem directly but maybe i can build a workaround. 

Thanks a lot for your informations

**Comment by jboxberger** [comment]
@iConMayrhofer when you got your backup running, and before the next upograde. You should rename the "12" folder from the psql data folder in something like "12.backup". I am not sure about that but i think this may cause problems when you try to install a working psql 12 package someday.

**Comment by iConMayrhofer** [comment]
I think i would have time to test the 13.0.3 Version a second time. If that helps.

**Comment by iConMayrhofer** [comment]
So this are my logs from 13.0.3. Thats all.

[synology_gitlab_postgresql.log](https://github.com/jboxberger/synology-gitlab/files/4748792/synology_gitlab_postgresql.log)
[synology_gitlab.log](https://github.com/jboxberger/synology-gitlab/files/4748793/synology_gitlab.log)

**Comment by jboxberger** [comment]
ok this one says 

```
2020-06-08 21:35:22	stdout	2020-06-08 21:35:22.354 UTC [167] STATEMENT:  CREATE DATABASE "gitlab" ENCODING = 'unicode'
2020-06-08 21:35:22	stdout	2020-06-08 21:35:22.354 UTC [167] ERROR:  database "gitlab" already exists
```

could you please check wether the folder /volume1/docker/gitlab/postgressql/12 (something like that) exists? This seems to prevent the migration von beening executed.

**Comment by iConMayrhofer** [comment]
Yeah, there is a 12 Folder

**Comment by jboxberger** [comment]
I assume you need to rename it before the install... there is the broken "gitlab" database from the previous installation. i am actually not on my dev system to check this, but i am pretty sure that this is the issue why migration says that the database already exist.

**Comment by iConMayrhofer** [comment]
I deleted the 12 Folder bevor installation. But as Version 13.0.3 uses Postgre 12, it installed it.
(If it isn't clear I'm talking about the original 13.0.3. I will test the 13.0.3 with Postgre10 Version now.)

**Comment by iConMayrhofer** [comment]
Version 13.0.3 with Postgre 10 works.
But an Update from there to 13.0.5 doesn't. Migration doesn't seem to work.
Here are my logs
[synology_gitlab_postgresql.log](https://github.com/jboxberger/synology-gitlab/files/4750192/synology_gitlab_postgresql.log)
[synology_gitlab.log](https://github.com/jboxberger/synology-gitlab/files/4750193/synology_gitlab.log)

**Comment by jboxberger** [comment]
@iConMayrhofer: thanks a lot for your testes. Yes you're absolutely right. the migration from postgres 10 to 12 fails because of your data set. 

I can not really do something about it since i can not reproduce the problem with my data set and because of security concerns i can not ask for your data set to test around.

This is The Problem (from your synology_gitlab_postgresql.log):
```
2020-06-09 05:37:16	stdout	2020-06-09 05:37:16.519 UTC [143] ERROR:  duplicate key value violates unique constraint "application_settings_pkey"
```
This is Maybe the Solution: https://gitlab.com/gitlab-org/gitlab-foss/-/issues/31976#note_35234678

Since the log is not from the beginning i can only guess that the "application_settings_pkey" is the error, maybe the "application_settings_pkey" is one of the errors, but once it is solved we can watch for the next ones and solve them one by one.

Another Problem from the gitlab container is:
```2020-06-09 05:42:00	stdout	ActiveRecord::StatementInvalid: PG::DuplicateTable: ERROR:  relation "abuse_reports" already exists```
I am not sure why this error happen. This may happen because the files from a previous migration attempt (/volume1/docker/gitlab/postgresql/12) were not deleted before the upgrade or the error is cause by another issue which is not in the log.

I've build a version for you with postgres 11.8, maybe this one can migrate your dateset correctly, but i do not really believe this because this error is a hard database error. I don't know if the constariant application_settings_pkey is new and why it did not lead t to an error previously, but for now it is what it is and has to be solved.
[synology-gitlab_synology-gitlab-stock-aio-11.11.0-0053.spk](https://send.firefox.com/download/2df3b315be4bec8d/#oAWEJGloI9kUO1bljBVqDg)

I try to test around to get some more information's. If you get any new errors please post them here, they are very helpful to get the problem isolated.

Kind Regards

**Comment by jboxberger** [comment]
I have an idea for a workaround.
1) backup gitlab with psql 10
2) install gitlab with same version but psql 12
3) restore backup from gitlab
4) update to higher versions

If you want to give it a try. Please tell me you current version so i can provide you the same version with psql 12.

Kind Regards

**Comment by iConMayrhofer** [comment]
I'm on Gitlab 12.9.2-0055

**Comment by jboxberger** [comment]
Here you go: [synology-gitlab-stock-aio-12.9.2-12-0055.spk](https://send.firefox.com/download/393c8d9af8ff897a/#97x_nyzuUtGD8OTaCPoboQ)

**Comment by iConMayrhofer** [comment]
It worked. I didn't even need to restore my backup. I simply installed it over the old version and everything was there after the Update.

So i will try to install a 13.0.X Version.
Should i test the 13.0.5 Version or go with the 13.0.3?

**Comment by jboxberger** [comment]
Lets stick with the 13.0.3, i will upload the version with psql 12 in a moment for you.

**Comment by jboxberger** [comment]
Download: [synology-gitlab-stock-aio-13.0.3-12-0055.spk](https://send.firefox.com/download/ecc9a67f39532a3e/#gkaiVxWq_Pe14mOiah-ZkA)

**Comment by helmut-steiner** [comment]
Thank you guys for getting to the bottom of this issue!! I am looking forward to testing this on the weekend. :)

**Comment by jboxberger** [comment]
@helmut-steiner before you start, could you please inspect your psql table application_settings for duplicate keys. i need to find a solution, a workaround is great but i can not publish a new release that fails to migrate. I am glad that you two guys have made backups and are familiar with backup, restoring and reinstalling but not everyone using this package has this skills even if its documented.

@iConMayrhofer @helmut-steiner : maybe you can dump your "faulty" application_settings for me (of course replacing the sensitive data) so i can build a check or a fix before installation or something like that.

**Comment by iConMayrhofer** [comment]
If you tell me the best way to dump my application_settins, i can do it.

**Comment by jboxberger** [comment]
Ok, a am at work at the moment, tonight i will build you a command to dump it to a sql file. :thumbsup:

**Comment by iConMayrhofer** [comment]
Ok, Thanks for your help and this repo. Good Work 👍

**Comment by helmut-steiner** [comment]
I ran into another issue after reinstalling version 12.9.2 and restoring the backup:
The runners pages in GitLab are not accessible anymore and throw a 500 error. The backup process doesn't save the secrets.json file and I don't have it in my hyper backup as it is buried somewhere in the NAS folder structure outside of the visible volume.
So long story short: I lost the secrets.json file after restoring the backup and cannot access the admin pages for my runners. There is a way to reset the runners but I don't know how to access them in the Synology environment: https://docs.gitlab.com/ee/raketasks/backup_restore.html#reset-runner-registration-tokens 
There exists a stackoverflow question about this as well:
https://stackoverflow.com/questions/54216933/internal-server-error-500-while-accessing-gitlab-admin-runners

How can I access the right console in the NAS environment? Do you know the commands for resetting the runners? Any help is much appreciated!

@jboxberger After solving this, the steps to backup and restore the secrets file should be added to the repo description. What do you think?

**Comment by jboxberger** [comment]
#### start dbconsole:
```
sudo docker exec -it synology_gitlab bash -c "sudo /home/git/gitlab/bin/rails dbconsole -e production"
```

password is in your environment vars (default: gitlab_pass):
![grafik](https://user-images.githubusercontent.com/1545845/84184746-346f2a80-aa8e-11ea-8304-6e279ca7bbd6.png)

#### start gitlab shell:
```
sudo docker exec -it synology_gitlab bash -c "sudo -u git -H bundle exec rails console -e production"
```
#### bash into container: 
```
sudo /usr/local/bin/docker exec -it synology_gitlab bash
```

did this help?

> @jboxberger After solving this, the steps to backup and restore the secrets file should be added to the repo description. What do you think?

of course this should!

**Comment by jboxberger** [comment]
@iConMayrhofer here i got some basic commands for you and as promissed the command for backing up the table. 

#### dump table application_settings
```sudo docker exec -i synology_gitlab_postgresql bash -c "sudo -u postgres pg_dump --table public.application_settings gitlab" > /volume1/docker/application_settings.sql```

#### dump database
```sudo docker exec -i synology_gitlab_postgresql bash -c "sudo -u postgres pg_dump gitlab" > /volume1/docker/outfile.sql```

#### restore database
```cat /volume1/docker/outfile.sql | sudo docker exec -i synology_gitlab_postgresql bash -c "sudo -u postgres psql gitlab"```

#### bash into postgresql container
```sudo docker exec -it synology_gitlab_postgresql bash```

#### psql console
```sudo -u postgres psql```

#### use gitlab schema
```\c gitlab```
#### show tables
```\dt```

#### quit
```\q```

**Comment by helmut-steiner** [comment]
The dbconsole line did the trick. Thanks Juri! 
fyi: I had to wait a couple of minutes on my old NAS till the console started up. The first two times I tried I thought it hung somewhere. Once I got my runners set up again I'll dump those application settings.

**Comment by helmut-steiner** [comment]
I couldn't see any duplicate keys in the settings table but maybe the dump helps you in some way:
https://nas.helmutsteiner.net:65502/sharing/ABLN9BOZ8
Link is valid for 7 days.

**Comment by iConMayrhofer** [comment]
https://icongmbhsrl-my.sharepoint.com/:u:/g/personal/k_mayrhofer_icon_bz_it/ERl4G-z-T3hLj5ukTquXu3EBUJIegZrhl3WpFep5SOhcbw?e=i82rs8
Link is valid for 7 days

**Comment by jboxberger** [comment]
OK thank you both for providing the example files but i could not locate the exact problem.

@iConMayrhofer: your db version was from 13.0.3 so i could not test the migration from 12.9.2 to 13.0.3

@helmut-steiner: my migration went fine with your application_settings. no errors at all.

So is still need a full log :-(.

Do you guys get the same error with the PSQL 11 version?
[synology-gitlab-stock-aio-13.0.3-11-0055.spk](https://send.firefox.com/download/f8a484ce92fc1e4a/#bJZLXeJD61sPwIuxQOUOuQ)

**Comment by iConMayrhofer** [comment]
@jboxberger 
Oh sorry, I have already deleted everything from version 12.9.2

> Do you guys get the same error with the PSQL 11 version?
synology-gitlab-stock-aio-13.0.3-11-0055.spk

Sorry for my late reply, if you'd still like to test it, I could try it eventually. I just don't know when

**Comment by helmut-steiner** [comment]
Hey Juri,
I still didn't get the chance to try the PSQL 11 version, but as everything worked on your end with v12 I will try the migration on my second NAS device. I cannot risk the downtime at the moment on my production environment as we are in the middle of intense testing and bugfixing for an upcoming release.
Sorry for the inconvenience!

**Comment by caco3** [comment]
Is there progress on the issue? Due to some issues with `12.9`, I would like to upgrade.

**Comment by helmut-steiner** [comment]
Hey,
I tried the upgrade one more time yesterday on our production environment NAS (which is 7y old by now) and it failed again with the same issue. Restoring the backup from there on my newer NAS and upgrading to the 13.0.3 version there: No issues! It seems like the old NAS' CPU is to slow to handle the database migration process in time or it might be another resource issue leading to this. 
I backed up the 13.0.3 version from my test NAS and deployed it onto a fresh 13.0.3 installation on the production NAS. Everything is back to normal now. I guess if you got a newer device the upgrade works just fine. We are ordering a new production device next week to be on the safe side for the future.

IMPORTANT NOTES:
The upgrade process overwrites the secret keys in the docker container configuration. (Sidenote at @jboxberger: They should not be overwritten during the upgrade process.) Be sure to save the secrets before upgrading if you want to restore a backup with configured runners!!! You can also copy the secrets.yml and gitlab.yml files to the backup folder using following commands:
```
sudo /usr/local/bin/docker exec -it synology_gitlab bash
cp config/gitlab.yml ../data/backups/
cp config/secrets.yml ../data/backups/
```
or copy the whole config folder to your backup folder and restore it after restoring the backup.

Kind regards,
Helmut

**Comment by jboxberger** [comment]
Hi Helmut, 

thanks for your backup information. I've spent the whole afternoon to get the GitLab 13 working. But there is something wrong with the docker image. Although the migration seems to work sometimes from 12.9.2 to 13.0.3, the clean installation and migration to higher versions fails. I am not sure how to solve this without change everything and keep the compatibility. 

The logical step would be a change to the official giltab-ce and postgres image. This could be curse or blessing.

I will keep you up to date.

**Comment by jboxberger** [comment]
I am going for a new package based on the official gitlab-ce postgres and redis image. This is a lot of work and i can't promise to get this done. I need to learn setup the gitlab-ce properly and combine this with the other images. I also aim to get this done with the official Synology toolkit to get this package signed. If this works out i will be able (hopefully) to get this package published in a repo so you can get auto updates.

wish me no luck, but patience and time :-)

**Comment by helmut-steiner** [comment]
This is good news. I thought about switching to the original packet myself a couple of times already.

Do you need help?

**Comment by iConMayrhofer** [comment]
I also think this will be the best solution. The update never worked on our two year old NAS. So it should be a fast enough system.

I can also try to help if needed

**Comment by helmut-steiner** [comment]
For anyone interested:
Juri updated the build script and upgrading to 13.3.4 works, but beware - fresh installs don't. The 13.3.4 package can be downloaded here for the next two weeks: https://we.tl/t-2UjBRmFiN6
Please try the upgrade on a non production system beforehand! Always make backups!! No guarantees are given. Also I had to stop and restart the package after the update to get it running on my production system.

I am really looking forward to the new package with the original gitlab-ce images @jboxberger !! :D

**Comment by jboxberger** [comment]
Hello guys, this took me half a day to figure it out but now i have at least a diry "workaround".

There were two issues, one were the missing "btree_gist" and the second is the missing "ghost" column in the users table. 

Today i solved the issue with the btree_gist together with @helmut-steiner.

But the missing column is still an issue. [The Column has been removed with GitLab 13.4.0](https://gitlab.com/gitlab-org/gitlab/-/blob/master/CHANGELOG.md). I traced the missing column down to 13.0.6 so with the newer 13.2.6 the column is missing but the GitLab Codes seems still to use it. [There are same issues out there](https://gitlab.com/gitlab-org/omnibus-gitlab/-/issues/5627).

As a Workaround we need to add the column.

**Auto ADD:**
`sudo /usr/local/bin/docker exec --user=postgres -it synology_gitlab_postgresql bash -c "psql -d gitlab -c 'ALTER TABLE users ADD COLUMN IF NOT EXISTS ghost BOOLEAN'"`

**Auto REMOVE (whene not needed anymore):**
`sudo /usr/local/bin/docker exec --user=postgres -it synology_gitlab_postgresql bash -c "psql -d gitlab -c 'ALTER TABLE users DROP COLUMN ghost'"`

**Manual:**
```
Docker App => Terminal => Create => bash
sudo -u postgres psql -d gitlab
ALTER TABLE users ADD COLUMN IF NOT EXISTS ghost BOOLEAN;
\q
```

@helmut-steiner could you give it a try. Just pull the current build script and add the column when you get the 500 on you login screen.

Unfortunately this will solve this specific issue, but there could be more of them because i could not find out why the column is missing but is still in use by the code. There seems to be a mismatch in code an schema version and there could be more columns missing or may have the wrong data type.

**Comment by helmut-steiner** [comment]
Hey Juri,
If you are a signed in user the issue doesn't occur. That's why I didn't see it during my first tests as I still had my session cookie.
I tried it on my test system and the workaround (auto add column) did the trick. The login screen works as expected.

**Comment by jboxberger** [comment]
Hi Helmut,

could you please do some more testing with your data set and see if you find something broken. If you do not find further issues i am thinking about to make a "experimental" release so everybody who feels save with backup and restore can give it a try.

Kind Regards

**Comment by helmut-steiner** [comment]
Yeah, I will play around a bit tomorrow and let you know. :)

**Comment by helmut-steiner** [comment]
Okay, I ran further tests:

- Login works
- Checkout/pull/push works
- Admin area works as far as I tested it
- Resetting runners in the DB works (if you forgot your secrets or didn't back them up...)
- Creating/closing/modifiying/assigning issues works
- Basic project/repository sites are accessible (graph, commits, wiki, analytics, settings, members...)

So far things look good. :)

**Comment by iConMayrhofer** [comment]
Good work guys. I will help to test if we have an experimental release

**Comment by helmut-steiner** [comment]
Version 13.4.2 has been released by sameersbn. @jboxberger Have you tried if the issues still persist with that version? I can't build it with the script.

**Comment by jboxberger** [comment]
Hi @helmut-steiner, i could build it without any issues. What was the error? I've tested the 13.4.3  and the problem still persist. But i published a experimental release. 

https://github.com/jboxberger/synology-gitlab/releases

I've also discovered a non breaking error during the update from 12.9.2 to 13.4.3 but i could no see any effects. Same problem here a missing column... :-1: 

```
2020-10-09 13:01:37.004 UTC [143] ERROR:  column "default" of relation "services" does not exist at character 200

2020-10-09 13:01:37.004 UTC [143] STATEMENT:  WITH created_records AS (

	  INSERT INTO services (project_id, "active","properties","type","template","push_events","issues_events","merge_requests_events","tag_push_events","note_events","category","default","wiki_page_events","pipeline_events","confidential_issues_events","commit_events","job_events","confidential_note_events", created_at, updated_at)

	  SELECT "projects"."id", true,'{}','PrometheusService',false,true,true,true,true,true,'monitoring',false,true,true,true,true,true,true, TIMEZONE('UTC', NOW()) as created_at, TIMEZONE('UTC', NOW()) as updated_at FROM "projects" LEFT JOIN services ON services.project_id = projects.id AND services.project_id BETWEEN 1 AND 1

	                    AND services.type = 'PrometheusService' INNER JOIN cluster_groups ON cluster_groups.group_id = projects.namespace_id INNER JOIN clusters_applications_prometheus ON clusters_applications_prometheus.cluster_id = cluster_groups.cluster_id

	                      AND clusters_applications_prometheus.status IN (3, 5) WHERE (projects.id BETWEEN 1 AND 1 AND services.id IS NULL)

	  RETURNING *

	)

	SELECT COUNT(*) as number_of_created_records

	FROM created_records

```

**Comment by helmut-steiner** [comment]
That's the thing: I don't even get an error. It just doesn't do all the things it is supposed to. It doesn't download the images and just builds the final file which is only 560kb big.
```
============================================================================================================
 Please provide the full qualified docker container name from dockerhub.com or version number
============================================================================================================
GitLab [default sameersbn/gitlab:13.4.3]: 13.4.3
Postgres [default sameersbn/postgresql:11-20200524]:
Redis [default redis:5.0.9]:
export image sameersbn/gitlab:13.4.3
  100 %               240 B / 203 B = 1.182
export image sameersbn/postgresql:11-20200524
  100 %               240 B / 203 B = 1.182
export image redis:5.0.9
  100 %               240 B / 203 B = 1.182
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  540k  100  540k    0     0   861k      0 --:--:-- --:--:-- --:--:--  859k
```

Btw: Are we using postgresql 11 or 12 now?!

**Comment by iConMayrhofer** [comment]
The 13.4.3 Version doesn't work for me. After installation, I get a 502 error. And if I add the ghost column, it works, but my credentials don't. Also, my custom login screen is lost.

**Comment by helmut-steiner** [comment]
@iConMayrhofer Sounds like the problem I had with my old NAS where the CPU was too slow and the DB migration got into a timeout. Which NAS are you using?

**Comment by iConMayrhofer** [comment]
DS1517+

**Comment by jboxberger** [comment]
Hello @iConMayrhofer, 

the advice of Helmut might be correct. As far as i understood the Gitlab DB Upgrade procedure happens incremental. So by restarting this will continue where it stopped. But not if there were a other error in between. You need to check the log files and search for errors. Please post them here, maybe we  can help.

![grafik](https://user-images.githubusercontent.com/1545845/96179729-9d314c80-0f31-11eb-98e6-07a5f7a6488c.png)

Kind Regards

**Comment by jboxberger** [comment]
I've got good news. I got the Synology toolkit running and i got the gitlab-ce running, i like this solution way more than the synology-gitlab approach. I will aim for no need for spk, but will offer a very thin installer for beginners.

@helmut-steiner : do you want to give it a try? This can run absolutely parallel to the current installation. You just need to create the folder where the Gitlab files should be located.

Please rename the gitlab-ce-13.4.3-ce.0.json.txt => gitlab-ce-13.4.3-ce.0.json
[gitlab-ce-13.4.3-ce.0.json.txt](https://github.com/jboxberger/synology-gitlab/files/5387616/gitlab-ce-13.4.3-ce.0.json.txt)

create:
/docker/gitlab-ce/data
/docker/gitlab-ce/logs
/docker/gitlab-ce/config

![grafik](https://user-images.githubusercontent.com/1545845/96181833-e46d0c80-0f34-11eb-8888-cd5ea1bb573f.png)

Pull docker image:
![grafik](https://user-images.githubusercontent.com/1545845/96180876-69572680-0f33-11eb-8acf-5d77082ea37b.png)

Import json in docker:
![grafik](https://user-images.githubusercontent.com/1545845/96180633-049bcc00-0f33-11eb-88a9-05422245c96b.png)

![grafik](https://user-images.githubusercontent.com/1545845/96180686-17160580-0f33-11eb-87bd-f657ad403058.png)

start container:
![grafik](https://user-images.githubusercontent.com/1545845/96181984-1c744f80-0f35-11eb-876a-d52a2caf7bca.png)

After the container booted up (first time takes a while, ready when about 2GB RAM usage) you can access it via 
http://your-host-or-ip:30080

The great about this solution, only one container, exposed config and logfiles (be careful with the permissions). 

Could you please test if you can migrate your data into gitlab-ce by simply export it from your current installation and import them into the gitlab-ce?

Kind Reagrds

**Comment by caco3** [comment]
> I've got good news. I got the Synology toolkit running and i got the gitlab-ce running, i like this solution way more than the synology-gitlab approach. I will aim for no need for spk, but will offer a very thin installer for beginners.

I think this is the same approach as documented by Matthias Lohr on his [website](https://mlohr.com/gitlab-on-a-diskstation/).
I switched myself to that approach due to the issue here. As far as I can see it works fine, how ever the CPU load is higher than with your spk-approach, see https://mlohr.com/gitlab-on-a-diskstation/#comment-33301

@jboxberger and @helmut-steiner can you confirm this?

**Comment by helmut-steiner** [comment]
@jboxberger I will give it a try!

**Comment by fsutter** [comment]
@jboxberger how to import into gitlab-ce?
Thanks

**Comment by helmut-steiner** [comment]
@fsutter 

1) Copy your backup file to docker\gitlab-ce\data\backups.
2) From the official Gitlab documentation:

```
# Stop the processes that are connected to the database
sudo docker exec -it <name of container> gitlab-ctl stop unicorn
sudo docker exec -it <name of container> gitlab-ctl stop puma
sudo docker exec -it <name of container> gitlab-ctl stop sidekiq

# Verify that the processes are all down before continuing
sudo docker exec -it <name of container> gitlab-ctl status

# Run the restore
sudo docker exec -it <name of container> gitlab-backup restore BACKUP=11493107454_2018_04_25_10.6.4-ce

# Restart the GitLab container
sudo docker restart <name of container>
```

**Comment by fsutter** [comment]
@helmut-steiner 
I've also found these instructions but I got this error message:
No backups found in /var/opt/gitlab/backups
Please make sure that file name ends with _gitlab_backup.tar

How do I change the default backup directory?
I haven't the /var/opt/gitlab directory on my synology

Thanks

**Comment by helmut-steiner** [comment]
@fsutter Copy the backup file in the data\backups folder in your gitlab-ce folder. 

@jboxberger I installed the gitlab-ce 13.0.3 version and restored my production backup there. It works fine. Now I am looking into how to get HTTPS going as letsencrypt doesn't have access to the special ports. I am thinking of copying the letsencrypt certs from the NAS to the gitlab-ce installation. Thoughts on that for a better way as this has to be done manually after each renewal?

**Comment by fsutter** [comment]
Hi, 
it worked also for me.
Now I have to figure out how to "configure" it.
I'm using an external reverse proxy and with the synology package I used GITLAB_HTTPS, GITLAB_HOST and GITLAB_PORT environment variables to specify how to access from external.
I've also used GITLAB_BACKUP_EXPIRY to rotate backups.

Frédéric

**Comment by helmut-steiner** [comment]
Hey there,
I wrote a script to copy the certificate and key files into the ssl folder under config:
1) Create a "ssl" folder in gitlab-ce/config for user root:root with access rights 755
2) Create a securemygitlab.sh file with following content and replace the gitlab.yourdomain.com with your gitlab domain:
```
#!/bin/bash
echo Copying TLS certificates to gitlab...
cd /usr/syno/etc/certificate/system/default
echo Copy new gitlab.crt
rm /volume1/docker/gitlab-ce/config/ssl/gitlab.yourdomain.com.crt
cat fullchain.pem cert.pem > /volume1/docker/gitlab-ce/config/ssl/gitlab.yourdomain.com.crt
echo Copy new gitlab.key
rm /volume1/docker/gitlab-ce/config/ssl/gitlab.yourdomain.com.key
cp privkey.pem /volume1/docker/gitlab-ce/config/ssl/gitlab.yourdomain.com.key
echo Done.
```
3) sudo ./securemygitlab.sh

All set for https! :D

Be sure to set following variables in gitlab.rb 
`external_url` 'https://gitlab.yourdomain.com'

Also set letsencrypt to false:
`letsencrypt['enable']` = false

**Comment by helmut-steiner** [comment]
You can also use the script in the task scheduler of the NAS to execute every couple of days so your gitlab instance will always have an up-to-date certificate. :)

**Comment by helmut-steiner** [comment]
Aaaand finally: Update from gitlab-ce 13.0.3 to 13.4.4 worked. :)
I will play around a bit but except the higher CPU usage so far nothing speaks against production usage.

**Comment by fsutter** [comment]
@helmut-steiner I tried to change external_url but after gitlab is no more responding :(
I tries either by passing GITLAB_OMNIBUS_CONFIG in the .json file or by modifying gitlab.rb, same problem.

Thanks

**Comment by helmut-steiner** [comment]
@fsutter
Did you restart the docker container after changing the URL?
Is your Gitlab URL different than your NAS URL? If so the above script doesn't work.

**Comment by fsutter** [comment]
@helmut-steiner Yes, it's precisly when restarting that gilab never respond even after 20 minutes.
If I put the line in comment it works again.
I haven't tried the script yet.

ragards,

**Comment by helmut-steiner** [comment]
@fsutter If your URL contains https then you have to add the certificate and key file before starting up the container. Otherwise it won't work.

**Comment by fsutter** [comment]
@helmut-steiner I'm using a resverse proxy that makes the http->https "translation".
How do I deal with that?

**Comment by helmut-steiner** [comment]
@fsutter Sorry I haven't looked into that yet. I might have time on the weekend to experiment with setting up the reverse proxy. There is some information on this blog posts and in the comments on how to get reverse proxy working: 
https://mlohr.com/gitlab-on-a-diskstation/

**Comment by jboxberger** [comment]
@fsutter 
I am not into reverse proxies, but could you please have a look here. Maybe this helps. I am also very interested in the solution, if you get this running please post how, i will add it later for a wiki.

https://www.itsfullofstars.de/2019/06/gitlab-behind-a-reverse-proxy/

`external_url 'https://gitlab.itsfullofstars.de'`

`gitlab-ctl reconfigure`

**Comment by jboxberger** [comment]
@caco3: Thank you for your link. I haven't knew that it's that easy since the sameersbn image is really complicated i could not really get into it. I started the project around Gitlab 8.x.x, since then some things changed an i grew with the project. I started as an absolute novice in Gitlab and Docker so i took something working and improved it, but now i am a bit further. 

Once everything is up and running a will make a thin installer for beginners. This thin installer aims for people who are new to Synology, Docker, Gitlab and just want to get it running. I want it to be fully compatible to the Manual configuration so you can later simply switch.

Kind Regars

**Comment by helmut-steiner** [comment]
Okay. I tried it and the reverse proxy works flawlessly with my setup. Here are my settings:
![Screenshot 2020-10-16 142307](https://user-images.githubusercontent.com/8708898/96258116-d44c4000-0fbb-11eb-865b-4a1fed6b3dfe.jpg)
![Screenshot 2020-10-16 142414](https://user-images.githubusercontent.com/8708898/96258122-d4e4d680-0fbb-11eb-87da-92b69db6ac60.jpg)

**Comment by fsutter** [comment]
Hi,

it works by adding this line to the .json:
      {
         "key" : "GITLAB_OMNIBUS_CONFIG",
         "value" : "external_url 'https://my.url.com'; nginx['listen_port'] = 80; nginx['listen_https'] = false"
      }

Thanks for the support and kind regards,

Frédéric

**Comment by helmut-steiner** [comment]
> Hi,
> 
> it works by adding this line to the .json:
> {
> "key" : "GITLAB_OMNIBUS_CONFIG",
> "value" : "external_url 'https://my.url.com'; nginx['listen_port'] = 80; nginx['listen_https'] = false"
> }
> 
> Thanks for the support and kind regards,
> 
> Frédéric

Ah, so you are proxying from https to an internal http server. Yeah. I put the requests right through - that's why I need the certificates also inside the container.

**Comment by fsutter** [comment]
Yes, I'm using HAProxy on pfsense.

Other problem: I did a backup from 12.9.2 gitlab-synology, installed gitlab-ce, restored and everything worked fine (except Web IDE but that's another story).
Then I deleted the container without removing /docker/gitlab-ce share and tried gitlab-ce-latest but it doesn't work, container will not start :(

**Comment by helmut-steiner** [comment]
I got one more question as well:
How do I set the SSH port to show the outside port 30022 or the formerly used 30001 inside gitlab instead of 22 which is used internally?

If I set `gitlab_rails['gitlab_shell_ssh_port'] = 3022 ` it also uses the 30022 port internally. Well maybe I just change the environment variables... Thoughts on that?

**Comment by helmut-steiner** [comment]
> I got one more question as well:
> How do I set the SSH port to show the outside port 30022 or the formerly used 30001 inside gitlab instead of 22 which is used internally?
> 
> If I set `gitlab_rails['gitlab_shell_ssh_port'] = 3022 ` it also uses the 30022 port internally. Well maybe I just change the environment variables... Thoughts on that?

Never mind. setting the `gitlab_rails['gitlab_shell_ssh_port'] = 3022` works just fine. There was something wrong with my port forwarding. 🤦‍♂️

**Comment by helmut-steiner** [comment]
> Yes, I'm using HAProxy on pfsense.
> 
> Other problem: I did a backup from 12.9.2 gitlab-synology, installed gitlab-ce, restored and everything worked fine (except Web IDE but that's another story).
> Then I deleted the container without removing /docker/gitlab-ce share and tried gitlab-ce-latest but it doesn't work, container will not start :(

@fsutter Do you have any error log or error message?

**Comment by fsutter** [comment]
@helmut-steiner no sorry, due to my Web IDE issues, I changed strategy: I made an export of all my projects, installed gitlab-ce-latest and re imported all projects into it.
Now all works fine.

regards,

Frédéric

**Comment by iConMayrhofer** [comment]
Upgrade to the `gitlab-ce:13.4.4-ce.0` image worked. 👍

**Comment by iConMayrhofer** [comment]
I found out that in version 13.4.4 three of my repos were empty.
I could fix this by restoring to version 13.0.3 and then manually doing the hashed storage miration:
`sudo docker exec -it gitlab-ce gitlab-rake gitlab:storage:migrate_to_hashed`

After that I could update to 13.4.4 without any problems.
I think the problem is somewhere in a version higher than 13.3, because until then I could update without problems.

**Comment by caco3** [comment]
@jboxberger Thanks for all the time you put into this. Will you provide a new/fixed package in the near future?

**Comment by crux153** [comment]
<img width="845" alt="Screen Shot 2021-03-12 at 10 21 41 AM" src="https://user-images.githubusercontent.com/10908841/110877441-c4dbe380-831c-11eb-8bba-e4a553a98ca6.png">

It looks like there's new stock synology package with version 13.6.2.
[https://www.synology.com/en-global/dsm/packages/Docker-GitLab](https://www.synology.com/en-global/dsm/packages/Docker-GitLab)

I don't think that they will be updating it regularly, so I would be very grateful if this can be updated to [recent version](https://about.gitlab.com/releases/2021/02/22/gitlab-13-9-released/).

**Comment by jboxberger** [comment]
@crux153 

Hi thanks for your hint. I played a bit around and it look good. Although i need to test the migration path from the previous versions. If you have the ability to test without risking your production data, i can provide you a test version. 

Kind Regards

**Comment by jboxberger** [comment]
Hello my friends, thank you for your patience... i've did a whole rewind of the project and tested a dozen of versions (not joking > 20h in total) to get a stable upgrade path. Please give me a few more days to cleanup and upload the packages.

@crux153: since you're on the experimental package, you need specialized version to get a clean migration. I will test the migration path for you so you have as less risk as possible. Please do Backups and check if you can restore them anywhere in a test environment just to prevent bad surprises. Your upgrade path is the risky one but i will help as good as i can.

**Upgrade for 13.0.3-0055**: 13.0.3-0055 =>  13.0.3-0058 => 13.6.2-0068
**Upgrade for 13.4.3-0055**: 13.4.3-0055 =>  13.4.3-0058 => 13.6.2-0068

@caco3: yes they will the problem were the migration of PSQL in the sameersbn gitlab image, it looks like it is solved by now so i can go on regular.

Preview:
![image](https://user-images.githubusercontent.com/1545845/111090875-70f61800-8531-11eb-823b-530f310a30b6.png)

**Comment by crux153** [comment]
@jboxberger Thank you for your kind help :) I do have automatic backups so I can help you with experiments. Do you want me to try upgrading to stock 13.6.2 package, or wait until you provide modified package?

**Comment by jboxberger** [comment]
Puh, i am done guys. Everything is updated and tested as far as i could. Please feel free to update, but do not forget, **backup, backup , backup** (Warning: GitLab does not back up any configuration files, SSL certificates, or system files. You are highly advised to [read about storing configuration files](https://docs.gitlab.com/ee/raketasks/backup_restore.html#storing-configuration-files)).

@crux153: you need to update to [13.0.3-0058](https://github.com/jboxberger/synology-gitlab/releases/tag/13.0.3-0058) first. After that you can update to 13.6.2-0068 and after that you can go for the 13.9.3-0068.

Please give me feedback so i can sleep peacefully knowing everything went fine :-).

**Comment by crux153** [comment]
@jboxberger I tried upgrading `13.0.3-0055-experimental` to new `13.0.3-0058`, but unfortunately it failed. During the update, a warning appeared that synology_gitlab_postgresql docker container failed unexpectedly and the progress got stuck at 100%. The setup continued to run once I manually started the postgresql and redis container, but all gitlab data were lost after the installation.

Fortunately I was able to restore backup from 13.0.3-0055 on 13.0.3-0058 without any problem. All my data came back again. From there, upgrading `13.0.3-0058` to `13.6.2-0068` and `13.9.3-0068` went butter smooth :)

So for anyone using the *experimental* package, be sure to **backup your data** before upgrade.

**Comment by miralsoft** [comment]
@jboxberger Hello, first of all I would like to thank you for your fantastic work here and also for the support you give here.

I've been hanging around updating my Gitlab for a long time now. I wanted to proceed according to your update strategy and wanted to update from 12.0.3-0054 to 12.4.2-0054. Unfortunately, already after the update, an error is issued during the migration that it does not work. In the attachment I have attached the log file from the image. Here you can see that an error occurs during the migration.

Do you have any idea what is going wrong here, or how I might get the migration through?

Thank you in advance and best regards
Michael 

[synology_gitlab_Update_12_4_2_0054.zip](https://github.com/jboxberger/synology-gitlab/files/6150382/synology_gitlab_Update_12_4_2_0054.zip)

**Comment by jboxberger** [comment]
Hi,

ho problem. I assume you created a beackup before.

In the LOG i see:
2021-03-16 15:23:43,stdout,"ArgumentError: Multiple indexes found on releases columns [:project_id]. Specify an index name from index_releases_on_project_id, releases_project_id_idx

It look like there is an index which is breaking the migration script. Have you already tried this?
https://github.com/jboxberger/synology-gitlab/wiki#database-migration-failed-during-update-1225-to-1233

Here i how you connect and execute queries on your posgresql
https://github.com/jboxberger/synology-gitlab/wiki#executing-sql-queries

Kind Reagrds

**Comment by miralsoft** [comment]
Hi,
thanks a lot for your quick help. Of course I overlooked this need. After I made this change that you suggested, the update worked fine for me. 

So that others don't have to search as long as I do, I'll briefly summarize my path here:
1. Open the GitLab console via ssh:
sudo / usr / local / bin / docker exec --user = postgres -it synology_gitlab_postgresql bash -c "psql -d gitlab"
2. Remove index:
\ d snippets
DROP INDEX index_snippets_on_project_id;

I have a backup every day and it also ends up on an external disk via HyperBackup, so I shouldn't lose anything. Without the backup, I would have lost my days because, before you wrote this post, I unfortunately wanted to go straight to 9/12/X without taking the steps before and my GitLab no longer wanted to run afterwards. Ergo went back to the old version and restored it via backup and tried again. Now I'm really happy that I've slowly got the configurations where I wanted (including ssl with Let's Encrypt and mail function with Microsoft 365).

I'll keep updating and hope that I won't get any more errors.

**Comment by jboxberger** [comment]
You ran into a known issue, as far as i know the rest will work fine, but you never know and the migration path depends on your specific gitlab usage. If you use many CI/CD for example and there is a bug in the CI/CD migration, the migration for you will fail. But if you do not use CI/CD at all your migration will work fine.  :-)

GitLab itself is really stable, but the sameersbn image, which synology is using, is kind of tricky sometimes. But with backups and patience we got everything running yet :-).

I wish you success and happy coding!

PS: I added your commands to the wiki, thanks for supporting.

Kind Regards

**Comment by caco3** [comment]
@jboxberger Thank you very much for your hard work!

 Today I did an update from `12.9.2-0055` to `12.10.6-0058` and didn't see any issues.

Then I directly upgraded `12.10.6-0058` to `13.9.3-0068` (I unintentionally skipped `13.6.2-0068`).

Again I can not see any issues or missing data (but I didn't check every single repo).

**Comment by jboxberger** [comment]
Thanks for your feedback, i am happy to hear that everything went fine! so i will close this issue. If you have any other issues please feel free to open a new one.

Kind Regards
